Google.Api.Generator.Utils/SourceFileContext.cs

```diff
@@ -162,51 +130,179 @@ public TypeSyntax Type0(Typ typ)
+                        .Join(
+                            couldBeImportedTypes,
+                            seen => seen.Key,
+                            imported => imported.Key,
+                            (seen, imported) => (type: seen.Key, namespaces: new HashSet<string>(seen.Value.Concat(imported))))
+                        .Where(tNs => tNs.namespaces.Count > 1)
+                        .ToDictionary(tNs => tNs.type, tNs => tNs.namespaces);
+
+
+                    // We now have all the information we need to know which types and namespaces need aliasing.
+                    // This is what we'll generate:
+                    // Collection of aliased namespaces. Namespace -> alias.
+                    Dictionary<string, string> aliasedNamespaces = new Dictionary<string, string>();
+                    // Collection of unaliased namespaces.
+                    HashSet<string> unaliasedNamespaces = new HashSet<string>();
+                    // Colletion of types that need to be aliased. Type -> namespaces.
+                    Dictionary<string, HashSet<string>> aliasedTypes = new Dictionary<string, HashSet<string>>();
+
+                    // Let's aliase namespaces until we have no collisions.
+                    string topCollidingNamespace;
+                    while ((topCollidingNamespace = GetTopCollidingNamespace()) != null)
+                    {
+                        // First aliase the namespace.
+                        var alias = ImportAliased(topCollidingNamespace);
+                        // Now aliase all colliding types from that namespace
+                        foreach (var tNs in collidingTypes.Where(ctNs => ctNs.Value.Contains(topCollidingNamespace)).ToList())
+                        {
+                            AliaseType(tNs.Key, topCollidingNamespace);
+                        }
+                    }
 
-                private readonly HashSet<string> _importedClassNames;
+                    // We have no collisiongs now. We can safely import remaining namespaces.
+                    unaliasedNamespaces.UnionWith(mustImportNs);
 
-                // Namespace -> alias
-                private readonly Dictionary<string, string> _namespaceAliases = new Dictionary<string, string>();
+                    return new TypeAliaser(aliasedNamespaces, unaliasedNamespaces, aliasedTypes);
 
-                public IReadOnlyDictionary<string, string> NamespaceAliases => _namespaceAliases;
+                    // Find the namespace with more collisions, prioritizing the ones that must be imported.
+                    // This is in an attempt to minimize imports and aliased namespaces, but it won't guarantee it.
+                    string GetTopCollidingNamespace() =>
+                        collidingTypes
+                            .SelectMany(tNs => tNs.Value)
+                            .GroupBy(ns => ns)
+                            .OrderBy(nsg => maySkipImportNs.Contains(nsg.Key))
```

Comment:
I think I'd use

```csharp
.OrderByDescending(nsg => mustImportNs.Contains(nsg.Key))
```

... just because that matches the comment more closely.